server/src/main/java/org/elasticsearch/common/logging/HeaderWarning.java

```diff
@@ -48,7 +51,31 @@ public class HeaderWarning {
         "\\d{4} " + // 4-digit year
         "\\d{2}:\\d{2}:\\d{2} " + // (two-digit hour):(two-digit minute):(two-digit second)
         "GMT" + // GMT
-        "\")?"); // closing quote (optional, since an older version can still send a warn-date)
+        "\")?",// closing quote (optional, since an older version can still send a warn-date)
+        Pattern.DOTALL
+    ); // in order to parse new line inside the qdText
+
+    /**
+     * quoted-string is defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6
+     * quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE
+     * qdtext         = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
+     * obs-text       = %x80-FF
+     * <p>
+     * this was previously used in WARNING_HEADER_PATTERN, but can cause stackoverflow
+     * "\"((?:\t| |!|[\\x23-\\x5B]|[\\x5D-\\x7E]|[\\x80-\\xFF]|\\\\|\\\\\")*)\"( " + // quoted warning value, captured
+     * the individual chars from qdText can be validated using the set of chars
+     * the \\\\|\\\\\" (escaped '\' 0x20 and '"' 0x5c) which is used for quoted-pair has to be validated as strings
+     */
+    private static Set<Integer> qdTextChars = Stream.of(
+        IntStream.of(0x09),// HTAB
+        IntStream.of(0x20), // SPACE
+        IntStream.of(0x21), // !
+        // excluding 0x22 '"\"' which has to be escaped
+        IntStream.rangeClosed(0x23, 0x5B),// ascii #-[
+        // excluding 0x5c '\' which has to be escaped
+        IntStream.rangeClosed(0x5D, 0x7E),// ascii ]-~
+        IntStream.rangeClosed(0x80, 0xFF)// obs-text -bear in mind it contains 0x85 new line. Which requires DOT_ALL flag
+    ).flatMapToInt(i -> i).boxed().collect(Collectors.toSet());
```

Comment:
Maybe make this unmodifiable. Or collect to boolean[] or bitset?

Reply:
bitset sounds good. Might be worth remembering that this is using only when testing, so performance should not be an issue